staging/src/k8s.io/apiserver/pkg/authentication/request/x509/x509.go

```diff
@@ -281,13 +284,45 @@ var CommonNameUserConversion = UserConversionFunc(func(chain []*x509.Certificate
 	fp := sha256.Sum256(chain[0].Raw)
 	id := "X509SHA256=" + hex.EncodeToString(fp[:])
 
+	uid, err := parseUIDFromCert(chain[0])
+	if err != nil {
+		return nil, false, err
+	}
 	return &authenticator.Response{
 		User: &user.DefaultInfo{
 			Name:   chain[0].Subject.CommonName,
+			UID:    uid,
 			Groups: chain[0].Subject.Organization,
 			Extra: map[string][]string{
 				user.CredentialIDKey: {id},
 			},
 		},
 	}, true, nil
 })
+
+var uidOID = asn1.X509UID()
+
+func parseUIDFromCert(cert *x509.Certificate) (string, error) {
+	if !utilfeature.DefaultFeatureGate.Enabled(features.AllowParsingUserUIDFromCertAuth) {
+		return "", nil
+	}
+
+	uids := []string{}
+	for _, name := range cert.Subject.Names {
+		if !name.Type.Equal(uidOID) {
+			continue
+		}
+		uid, ok := name.Value.(string)
+		if !ok {
+			return "", fmt.Errorf("unable to parse UID into a string")
+		}
+		uids = append(uids, uid)
```

Comment:
@enj do you think we should require the uid be non-empty if specified in a subject name, or tolerate it being explicitly present and set to an empty string?

Reply:
I updated UID parsing to return an error for the empty string case. I don't think an empty string would every be desirable, so returning an error seems like the best user experience.

Note that I didn't add a test because it doesn't seem possible to pass in empty values using the `openssl` command. For example, calling `openssl` with `-subj "/kube_uid="` elides the oid from `Certificate.Subject.Names`. (whereas `openssl` with `-subj "/kube_uid=abc"` includes it)